code/datums/medal_awards.dm

```diff
@@ -366,3 +463,185 @@ GLOBAL_LIST_INIT(xeno_medals, list(XENO_SLAUGHTER_MEDAL, XENO_RESILIENCE_MEDAL,
+
+	for(var/datum/medal_recommendation/recommendation in GLOB.medal_recommendations)
+		var/recommendation_list = list()
+
+		recommendation_list["rank"] = recommendation.recipient_rank
+		recommendation_list["name"] = recommendation.recipient_name
+		recommendation_list["ref"] = REF(recommendation)
+		recommendation_list["recommender_name"] = recommendation.recommended_by_name
+		recommendation_list["reason"] = recommendation.reason
+		recommendation_list["recommender_rank"] = recommendation.recommended_by_rank
+
+		data["recommendations"] += list(recommendation_list)
+	return data
+
+/datum/ic_medal_panel/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+	var/mob/living/carbon/human/user = usr
+	var/obj/item/card/id/card = user.wear_id
+	if(!card)
+		to_chat(user, SPAN_WARNING("You must have an authenticated ID Card to award medals."))
+		return
+
+	if(!((card.paygrade in GLOB.co_paygrades) || (card.paygrade in GLOB.highcom_paygrades)))
+		to_chat(user, SPAN_WARNING("Only a Senior Officer can award medals!"))
+		return
+
+	if(!card.registered_ref)
+		user.visible_message("ERROR: ID card not registered in USCM registry. Potential medal fraud detected.")
+		return
+
+	var/real_owner_ref = card.registered_ref
+
+	if(real_owner_ref != WEAKREF(user))
+		user.visible_message("ERROR: ID card not registered for [user.real_name] in USCM registry. Potential medal fraud detected.")
+		return
+
+	switch(action)
+		if("grant_new_medal")
+			if(!user_locs[user])
+				return
+			if(give_medal_award(get_turf(user_locs[user])))
+				var/atom/user_machine = user_locs[user]
+				user_machine.visible_message(SPAN_NOTICE("[user_machine] prints a medal."))
+			. = TRUE
+
+		if("approve_medal")
+			var/recommendation_ref = params["ref"]
+			var/medal_type = params["medal_type"]
+			if(!(medal_type in GLOB.human_medals))
+				return
+			var/datum/medal_recommendation/recommendation = locate(recommendation_ref) in GLOB.medal_recommendations
+			if(!recommendation)
+				return
+			if(!user_locs[user])
+				return
+			if(recommendation.recipient_name == user.real_name)
+				return
+
+			var/choice = tgui_alert(user, "Would you like to change the medal text?", "Medal Citation", list("Yes", "No"))
+			var/medal_citation = recommendation.reason
+			if(choice == "Yes")
+				medal_citation = strip_html(input("What should the medal citation read?", "Medal Citation", null, null) as message|null, MAX_PAPER_MESSAGE_LEN)
+
+			var/confirm_choice = tgui_alert(user, "Are you sure you want to give a medal to [recommendation.recipient_name]?", "Medal Confirmation", list("Yes", "No"))
+			if(confirm_choice != "Yes")
+				return
+			var/atom/user_machine = user_locs[user]
+			if(give_medal_award_prefilled(get_turf(user_machine), user, recommendation.recipient_name, recommendation.recipient_rank, recommendation.recipient_ckey, medal_citation, medal_type, recommendation.recommended_by_ckey, recommendation.recommended_by_name))
```

Comment:
Using the turf makes it so `The floor prints a medal.` and it will print it out wherever you started the process - not where you end up. It should just be the tablet passed and then when printed it can print to the turf where the tablet is located.

Reply:
give_medal_award_prefilled takes the loc where the medal is placed.
The message is just below (without the use of get_turf)